contributors/design-proposals/hugepages.md

```diff
@@ -0,0 +1,281 @@
+# HugePages support in Kubernetes
+
+**Authors**
+* Derek Carr (@derekwaynecarr)
+* Seth Jennings (@sjenning)
+* Piotr Prokop (@PiotrProkop)
+
+**Status**: In progress
+
+## Abstract
+
+A proposal to enable applications running in a Kubernetes cluster to use huge
+pages.
+
+A pod may request a number of huge pages.  The `scheduler` is able to place the
+pod on a node that can satisfy that request.  The `kubelet` advertises an
+allocatable number of huge pages to support scheduling decisions. A pod may
+consume hugepages via `hugetlbfs` or `shmget`.  Huge pages are not
+overcommitted.
+
+## Motivation
+
+Memory is managed in blocks known as pages.  On most systems, a page is 4KB. 1MB
+of memory is equal to 256 pages; 1GB of memory is 256,000 pages, etc.  CPUs have
+a built-in memory management unit that manages a list of these pages in
+hardware. The Translation Lookaside Buffer (TLB) is a small hardware cache of
+virtual-to-physical page mappings.  If the virtual address passed in a hardware
+instruction can be found in the TLB, the mapping can be determined quickly.  If
+not, a TLB miss occurs, and the system falls back to slower, software based
+memory management.  This results in performance issues.  Since the hardware is
+fixed, the only way to alleviate the performance impact of a TLB miss is to
+increase the page size.
+
+A huge page is a memory page that is larger than 4KB.  On x86_64 architectures,
+there are two huge page sizes: 2MB and 1GB.  Sizes vary on other architectures,
+but the idea is the same.  In order to use huge pages, application must write
+code that is aware of them.  Transparent huge pages (THP) attempt to automate
+the management of huge pages without application knowledge, but they have
+limitations.  In particular, they are limited to 2MB page sizes.  
```

Comment:
Also, dynamic splitting of THPs into standard pages during certain conditions can result in undesirable application stalls.